pkgs/objective_c/test/ns_input_stream_test.dart

```diff
@@ -0,0 +1,273 @@
+          expect(inputStream.streamError, null);
+        });
+      });
+    });
+
+    group('large stream', () {
+      late NSInputStream inputStream;
+      final streamData = List.generate(100, (x) => List.filled(10000, x));
+      final testData = streamData.expand((x) => x).toList();
+
+      setUp(() {
+        inputStream = Stream.fromIterable(streamData).toNSInputStream();
+      });
+
+      test('partial read', () async {
+        inputStream.open();
+        final (count, data, hasBytesAvailable, status, error) =
+            await read(inputStream, 100000);
+        expect(count, lessThanOrEqualTo(100000));
+        expect(count, greaterThanOrEqualTo(1));
+        expect(data, testData.sublist(0, count));
+        expect(hasBytesAvailable, true);
+        expect(status, NSStreamStatus.NSStreamStatusOpen);
+        expect(error, isNull);
+      });
+
+      test('full read', () async {
+        inputStream.open();
+        final readData = <int>[];
+        while (true) {
+          final (count, data, hasBytesAvailable, status, error) =
+              await read(inputStream, Random.secure().nextInt(100000));
+
+          readData.addAll(data);
+
+          expect(error, isNull);
+          if (count == 0) {
+            expect(hasBytesAvailable, false);
+            expect(status, NSStreamStatus.NSStreamStatusAtEnd);
+            expect(readData, testData);
+            break;
+          }
+        }
+      });
+    });
+
+    test('error in stream', () async {
+      late NSInputStream inputStream;
+
+      inputStream = () async* {
+        yield [1, 2];
+        throw const FileSystemException('some exception message');
+      }()
+          .toNSInputStream();
+
+      inputStream.open();
+      final (count1, data1, hasBytesAvailable1, status1, error1) =
+          await read(inputStream, 10);
+      expect(count1, 2);
+      expect(data1, [1, 2]);
+      expect(hasBytesAvailable1, true);
+      expect(status1, NSStreamStatus.NSStreamStatusOpen);
+      expect(error1, isNull);
+
+      final (count2, _, hasBytesAvailable2, status2, error2) =
+          await read(inputStream, 10);
+      expect(count2, -1);
+      expect(hasBytesAvailable2, false);
+      expect(status2, NSStreamStatus.NSStreamStatusError);
+      expect(
+          error2,
+          isA<NSError>()
+              .having((e) => e.localizedDescription.toString(),
+                  'localizedDescription', contains('some exception message'))
+              .having((e) => e.domain.toString(), 'domain', 'DartError'));
+    });
+  });
+
+  group('delegate', () {
+    late NSInputStream inputStream;
+
+    setUp(() {
+      inputStream = Stream.fromIterable([
+        [1, 2, 3],
+      ]).toNSInputStream();
+    });
+
+    test('default delegate', () async {
+      expect(inputStream.delegate, inputStream);
+    });
+
+    test('assign to non-self', () async {
+      inputStream.delegate = [1, 2, 3].toNSData();
```

Comment:
I don't really understand how this delegate stuff works. Is this effectively swapping out the backing data stream of `inputStream` for this `NSData`? If so, the `NSData` contains the same data that the `inputStream` was constructed with, so the test could pass even if the delegate logic was broken. Can you change the `NSData`'s data?

Just for my own understanding, how can you use an `NSData` as a `NSStreamDelegate`? It doesn't seem to conform to that protocol.

Also, can you explain the control flow that you expect in this case? It's super hard to follow (mostly because I'm not very experienced with the delegate pattern). It seems that `NSStreamDelegate` only has one method, and your implementation of that method at input_stream_adapter.m:162 only ever logs throws an error, or does nothing. What code actually invokes the methods on the delegate?

Reply:
> I don't really understand how this delegate stuff works. Is this effectively swapping out the backing data stream of `inputStream` for this `NSData`? If so, the `NSData` contains the same data that the `inputStream` was constructed with, so the test could pass even if the delegate logic was broken. Can you change the `NSData`'s data?
> 
> Just for my own understanding, how can you use an `NSData` as a `NSStreamDelegate`? It doesn't seem to conform to that protocol.
> 
> Also, can you explain the control flow that you expect in this case? It's super hard to follow (mostly because I'm not very experienced with the delegate pattern). It seems that `NSStreamDelegate` only has one method, and your implementation of that method at input_stream_adapter.m:162 only ever logs throws an error, or does nothing. What code actually invokes the methods on the delegate?

I changed the delegate method to just passthrough to the delegate (if it is not `self`) and confirmed the behavior in tests. I also added `NSStreamDelegate` to the list of protocols that exist in `package:objective_c`.